vaadin-spring/src/test/java/com/vaadin/flow/spring/scopes/VaadinRouteScopeTest.java

```diff
@@ -150,4 +125,149 @@ public void destroySession_sessionAttributeIsCleanedAndDestructionCallbackIsCall
         verify(factory, times(2)).getObject();
     }
 
+    @Test
+    public void refresh_uiWithTheSameWindowName_beanInScopeIsDestroyedAfterRefresh() {
+        UI ui = mockUI();
+
+        UI anotherUI = makeAnotherUI(ui);
+
+        ExtendedClientDetails details = Mockito
+                .mock(ExtendedClientDetails.class);
+        Mockito.when(details.getWindowName()).thenReturn("bar");
+        ui.getInternals().setExtendedClientDetails(details);
+        anotherUI.getInternals().setExtendedClientDetails(details);
+
+        ui.getSession().addUI(ui);
+        ui.getSession().addUI(anotherUI);
+
+        mockServletContext(ui);
+
+        VaadinRouteScope scope = initScope(ui);
+
+        AtomicInteger count = new AtomicInteger();
+        scope.registerDestructionCallback("foo", () -> count.getAndIncrement());
+
+        scope.uiInit(new UIInitEvent(ui, ui.getSession().getService()));
+
+        navigateTo(ui, new NavigationTarget());
+
+        putObjectIntoScope(scope);
+
+        // close the first UI
+        ui.getSession().removeUI(ui);
+
+        // the bean is not removed since there is a "preserved" UI
+        Assert.assertEquals(0, count.get());
+
+        UI.setCurrent(anotherUI);
+
+        scope = initScope(anotherUI);
+
+        // the bean is not removed since there is a "preserved" UI
+        Assert.assertEquals(0, count.get());
+
+        navigateTo(anotherUI, new AnotherNavigationTarget());
+
+        // the bean is removed since navigation away from it's owner navigation target
+        Assert.assertEquals(1, count.get());
+    }
+
+    @Test
+    public void detachUI_uiWithDifferentWindowName_beanInScopeIsDestroyedwhenUIIsDetached() {
+        UI ui = mockUI();
+
+        UI anotherUI = makeAnotherUI(ui);
+
+        ExtendedClientDetails details = Mockito
+                .mock(ExtendedClientDetails.class);
+        Mockito.when(details.getWindowName()).thenReturn("bar");
+        ui.getInternals().setExtendedClientDetails(details);
+
+        ui.getSession().addUI(ui);
+        ui.getSession().addUI(anotherUI);
+
+        mockServletContext(ui);
+
+        VaadinRouteScope scope = initScope(ui);
+
+        AtomicInteger count = new AtomicInteger();
+        scope.registerDestructionCallback("foo", () -> count.getAndIncrement());
+
+        scope.uiInit(new UIInitEvent(ui, ui.getSession().getService()));
+
+        navigateTo(ui, new NavigationTarget());
+
+        putObjectIntoScope(scope);
+
+        // close the first UI
+        ui.getSession().removeUI(ui);
+
+        // the bean is not removed since there is a "preserved" UI
+        Assert.assertEquals(1, count.get());
+        count.set(0);
+
+        UI.setCurrent(anotherUI);
+
+        scope = initScope(anotherUI);
+
+        navigateTo(anotherUI, new AnotherNavigationTarget());
+
+        // the bean is not removed since there is a "preserved" UI
+        Assert.assertEquals(0, count.get());
```

Comment:
Hmm, why is no pre-destroy calls expected here? As anotherUI has no details and no window name, I would expect bean is destroyed, rather than UI is reset.

Reply:
It's worth to check several line above where `Assert.assertEquals(1, count.get());` is written.